### DISCUSSION TITLE
Printf for LoRaMAC 4.5.2

### DISCUSSION THREAD

**Discussion by meakash** [Lora-net/LoRaMac-node#1145] [Q&A] [answered]
Respected all,

 I have trying to use the STM32 L073 based Firmware for Periodic Lpp Uplink and although i am getting Join Requests and Join Accepts I am not getting any uplink data. So I tried to insert printf commands and include stdio.h file in the headers, but i am not getting any output on the console. 

Can please somebody help me with the Baud rate and the communication. I am currently using a Nucleo F303RE to program the .bin file to my Hardware.

**Comment by mluis1** [comment]
By default this project sets up the UART parameters with 921600 bits/s 8N1.
The choice for this bit rate is to avoid as much as possible to disturb the LoRaWAN protocol handling.
In addition one must be careful where in the code `printf` commands are added. A `printf` takes time to be executed which may disturb some tight LoRaWAN protocol timings.

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Comment by meakash** [comment, marked as answer]
Thank you so much @mluis1. I'll keep in mind to post it in the discussions tab next time.

**Reply by meakash** [reply]
What i figured out is that UART Board.c has the following Configuration line: 

UartMcuConfig( Uart_t *obj, UartMode_t mode, uint32_t baudrate, WordLength_t wordLength, StopBits_t stopBits, Parity_t parity, FlowCtrl_t flowCtrl ) 

But i dont see any such configuation in the main file so that i can edit the baudrate. Can I edit it, if yes then please where can find the appropriate file?

**Reply by mluis1** [reply]
If you search for `UartConfig` in your project you will be able to see where this function is called.

In this project the setup is done under each platform `board.c` file.

Lets assume you use a NucleoL073 platform.
In `src/board/NucleoL073/board.c` you will find a function named `BoardInitMcu` which has the following lines

https://github.com/Lora-net/LoRaMac-node/blob/05353ff70bec5349a26b23a02168accb0ea03ad9/src/boards/NucleoL073/board.c#L134-L138

**Reply by meakash** [reply]
Thank you so much.
So, if I want to use any other Peripheral, I need to implement the statements in board.c?

@mluis1  I want to use I2C and in ST's Stack (I cube LRWAN) I define all the I2C defin1tions in main.c.  Since i am not able to print logs due to some issues in my custom hardware, i am not able to figure it out.

**Reply by mluis1** [reply]
You can do as you wish.

We are currently preparing a tutorial on how to use this project and the tutorial application example is on how to use a BME280 I2C sensor. This tutorial is expected to be released in the coming months.

The main.c file for this example starts by initializing the MCU I2C peripheral as shown in below code excerpt.

As you can observe in this case we start by initializing the peripherals required for LoRaWAN operation by calling `BoardInitMcu`. Then we call the `I2cInit` function to initialize the MCU I2C peripheral.
We decided to do it like this for the tutorial purposes.  
However we could have put the `I2cInit` function call inside `BoardInitMcu` or `BoardInitPeriph` functions implementation. Again this is up to you to decide where to put it.

```c
/*!
 * I2C object used for BME280 sensor communication
 */
I2c_t I2c;

/*!
 * Main application entry point.
 */
int main( void )
{
    BoardInitMcu( );
    BoardInitPeriph( );

    // Initialize MCU I2C peripheral
    I2cInit( &I2c, I2C_1, I2C_SCL, I2C_SDA );

    // Initialize BME280 I2C sensor
    if( BME280Init( BME280_I2C_ADDR_SEC ) != LMN_STATUS_OK )
    {
        printf("[BME280] Unable to initialize BME280 I2C sensor...\n");
    }

```

As we are diverging from the initial subject of this discussion would you mind marking this topic as answered. In case you have other questions please create a new subject.

**Reply by meakash** [reply]
Thank you so much for the help with the answer. 

Regards,
Akash Rao
